cjworkbench/i18n/trans.py

```diff
@@ -0,0 +1,181 @@
+    
+    For example, you may have passed a string in the place of an integer.
+    """
+
+
+_translators = {}
+
+
+def _get_translations(locale):
+    """Returns a MessageTranslator object for the given locale.
+    
+    In order to parse the message catalogs only once per locale,
+    uses the _translators dict to store the created MessageTranslator for each locale.
+    """
+    if locale in _translators:
+        return _translators[locale]
+    _translators[locale] = MessageTranslator(locale)
+    return _translators[locale]
+
+
+def trans(locale, message_id, default=None, context=None, parameters={}, tags={}):
+    """Translates the given message ID to the given locale.
+    """
+    return _get_translations(locale).trans(
+        message_id, default, context, parameters, tags
+    )
+
+
+def restore_tags(message, tag_mapping):
+    """Replaces the HTML tags and attributes in a message.
+    
+    `tag_mapping` is a dict that for each tag name contains a new name `name` and new attributes `attrs`
+    
+    For each non-nested HTML tag found, searches `tag_mapping` for a replacement for its name and attributes.
+    If found, replaces with the ones found.
+    If not found, removes the tag but keeps the (escaped) contents.
+    
+    Nested HTML tags are removed, with their (escaped) contents kept
+    
+    Returns the new message
+    """
+    soup = BeautifulSoup(message, "html.parser")
+    bad = []
+    for child in soup.children:
+        if child.name:  # i.e. child is a tag
+            if child.name in tag_mapping:
+                child.attrs = tag_mapping[child.name].get("attrs", {})
+                child.name = tag_mapping[child.name]["tag"]
+                child.string = "".join(child.strings)
+            else:
+                bad.append(child)
+    for child in bad:
+        child.string = escape("".join(child.strings))
+        child.unwrap()
+    return str(soup)
+
+
+class MessageTranslator:
+    """Loads the message catalogs for a given locale and provides helper methods for message translation.
+    It uses plaintext messages with variables in `{var}` placeholders, e.g. in `"Hello {name}"`, `name` is a variable.
+    
+    Essentially, it is a wrapper for the combination of 
+    Babel (for loading PO files) and variable substitution and our HTML placeholder construct.
+    """
+
+    def __init__(self, locale):
+        self.locale = locale
+
+        try:
+            self.catalog = read_po(open(catalog_path(locale)))
+        except ValueError as error:
+            raise BadCatalogsError(
+                "The catalog for the given locale (%s) is badly formatted" % locale
+            ) from error
+        except Exception as error:
+            raise UnsupportedLocaleError(
+                "Can't load a catalog for the given locale (%s)" % locale
+            ) from error
+
+    def trans(self, message_id, default=None, context=None, parameters={}, tags={}):
```

Comment:
Can we call it `def trans_html(...)`? The output is valid HTML, and that's a key feature. It should be part of the function signature.